waspc/src/Wasp/Generator/Valid/Validator.hs

```diff
@@ -0,0 +1,73 @@
+module Wasp.Generator.Valid.Validator where
+
+import Data.Bifunctor (first)
+import Data.Functor (void)
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.List.NonEmpty as NE
+import Data.Maybe (catMaybes)
+import qualified Validation as V
+import Wasp.Util (indent)
+
+type Validation result = V.Validation (NonEmpty ValidationError) result
+
+data ValidationError = ValidationError
+  { message :: String,
+    fieldPath :: [String],
+    fileName :: Maybe String
+  }
+
+-- | Executes the given validator on the input and returns a list of validation errors. If there are
+-- no errors, returns an empty list.
+getValidationErrors :: (input -> Validation result) -> input -> [ValidationError]
+getValidationErrors validator =
+  maybe [] NE.toList . V.failureToMaybe . validator
+
+-- | Adds file name context to validation errors produced by the inner validator.
+withFileName :: String -> (a -> Validation result) -> (a -> Validation result)
+withFileName fileName' innerValidator =
+  mapErrors setFileName . innerValidator
+  where
+    setFileName err = err {fileName = Just fileName'}
+
+-- | Runs the validator on a specific field of the input, adding the field name to the error path.
+valueOfField :: (String, a -> b) -> (b -> Validation result) -> (a -> Validation result)
```

Comment:
Ok idea, now that I read the comment: you could call this `fieldValidator`.
Because it takes a validator, and then turns it into a field validator.